endpoint.go

```diff
@@ -235,11 +332,13 @@ func (ep *endpoint) Join(containerID string, options ...EndpointOption) error {
 
 	ep.Unlock()
 	defer func() {
-		ep.Lock()
 		if err != nil {
+			ep.Lock()
 			ep.container = nil
+			ep.Unlock()
+		} else {
+			ep.network.ctrlr.addEndpointToStore(ep)
```

Comment:
Also I am pretty sure you need to update the join status to endpoint store as an atomic update immediately after you construct the local container info. Otherwise there is a race if two containers in different hosts try to join. If addEndpointToStore fails atomic update(because of index being old) we need to re-update endpoint data from kv store which might reveal that a join already happened somewhere else and handle that case here accordingly.